File: src/main/scala/li/cil/oc/common/template/AssemblerTemplates.scala
package li.cil.oc.common.template

import java.lang.reflect.Method

import com.google.common.base.Strings
import li.cil.oc.OpenComputers
import li.cil.oc.api
import li.cil.oc.api.network.EnvironmentHost
import li.cil.oc.common.IMC
import li.cil.oc.common.Slot
import li.cil.oc.common.Tier
import li.cil.oc.util.ExtendedNBT._
import net.minecraft.inventory.IInventory
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.text.ITextComponent
import net.minecraftforge.common.util.Constants.NBT

import scala.collection.mutable
import scala.language.existentials

object AssemblerTemplates {
  val NoSlot = new Slot(Slot.None, Tier.None, None, None)

  private val templates = mutable.ArrayBuffer.empty[Template]

  private val templateFilters = mutable.ArrayBuffer.empty[Method]

  def add(template: NBTTagCompound): Unit = {
    val selector = IMC.getStaticMethod(template.getString("select"), classOf[ItemStack])
    val validator = IMC.getStaticMethod(template.getString("validate"), classOf[IInventory])
    val assembler = IMC.getStaticMethod(template.getString("assemble"), classOf[IInventory])
    val hostClass = tryGetHostClass(template.getString("hostClass"))
    val containerSlots = template.getTagList("containerSlots", NBT.TAG_COMPOUND).map((tag: NBTTagCompound) => parseSlot(tag, Some(Slot.Container), hostClass)).take(3).padTo(3, NoSlot).toArray
    val upgradeSlots = template.getTagList("upgradeSlots", NBT.TAG_COMPOUND).map((tag: NBTTagCompound) => parseSlot(tag, Some(Slot.Upgrade), hostClass)).take(9).padTo(9, NoSlot).toArray
    val componentSlots = template.getTagList("componentSlots", NBT.TAG_COMPOUND).map((tag: NBTTagCompound) => parseSlot(tag, None, hostClass)).take(9).padTo(9, NoSlot).toArray

    templates += new Template(selector, validator, assembler, containerSlots, upgradeSlots, componentSlots)
  }

  def addFilter(method: String): Unit = {
    templateFilters += IMC.getStaticMethod(method, classOf[ItemStack])
  }

  def select(stack: ItemStack) = {
    if (!stack.isEmpty && templateFilters.forall(IMC.tryInvokeStatic(_, stack)(true)))
      templates.find(_.select(stack))
    else
      None
  }

  class Template(val selector: Method,
                 val validator: Method,
                 val assembler: Method,
                 val containerSlots: Array[Slot],
                 val upgradeSlots: Array[Slot],
                 val componentSlots: Array[Slot]) {
    def select(stack: ItemStack) = IMC.tryInvokeStatic(selector, stack)(false)

    def validate(inventory: IInventory) = IMC.tryInvokeStatic(validator, inventory)(null: Array[AnyRef]) match {
      case Array(valid: java.lang.Boolean, progress: ITextComponent, warnings: Array[ITextComponent]) => (valid: Boolean, progress, warnings)
      case Array(valid: java.lang.Boolean, progress: ITextComponent) => (valid: Boolean, progress, Array.empty[ITextComponent])
      case Array(valid: java.lang.Boolean) => (valid: Boolean, null, Array.empty[ITextComponent])
      case _ => (false, null, Array.empty[ITextComponent])
    }

    def assemble(inventory: IInventory) = IMC.tryInvokeStatic(assembler, inventory)(null: Array[AnyRef]) match {
      case Array(stack: ItemStack, energy: java.lang.Number) => (stack, energy.doubleValue(): Double)
      case Array(stack: ItemStack) => (stack, 0.0)
      case _ => (null, 0.0)
    }
  }

  class Slot(val kind: String, val tier: Int, val validator: Option[Method], val hostClass: Option[Class[_ <: EnvironmentHost]]) {
    def validate(inventory: IInventory, slot: Int, stack: ItemStack) = validator match {
      case Some(method) => IMC.tryInvokeStatic(method, inventory, slot.underlying(), tier.underlying(), stack)(false)
      case _ => Option(hostClass.fold(api.Driver.driverFor(stack))(api.Driver.driverFor(stack, _))) match {
        case Some(driver) => try driver.slot(stack) == kind && driver.tier(stack) <= tier catch {
          case t: AbstractMethodError =>
            OpenComputers.log.warn(s"Error trying to query driver '${driver.getClass.getName}' for slot and/or tier information. Probably their fault. Yell at them before coming to OpenComputers for support. :P")
            false
        }
        case _ => false
      }
    }
  }

  private def parseSlot(nbt: NBTTagCompound, kindOverride: Option[String], hostClass: Option[Class[_ <: EnvironmentHost]]) = {
    val kind = kindOverride.getOrElse(if (nbt.hasKey("type")) nbt.getString("type") else Slot.None)
    val tier = if (nbt.hasKey("tier")) nbt.getInteger("tier") else Tier.Any
    val validator = if (nbt.hasKey("validate")) Option(IMC.getStaticMethod(nbt.getString("validate"), classOf[IInventory], classOf[Int], classOf[Int], classOf[ItemStack])) else None
    new Slot(kind, tier, validator, hostClass)
  }

  private def tryGetHostClass(name: String) =
    if (Strings.isNullOrEmpty(name)) None
    else Option(Class.forName(name).asSubclass(classOf[EnvironmentHost]))
}

File: src/main/scala/li/cil/oc/common/template/AssemblerTemplates.java
package li.cil.oc.common.template;

import li.cil.oc.api.Driver;
import li.cil.oc.api.driver.DriverItem;
import li.cil.oc.common.Slot;
import li.cil.oc.common.Tier;
import li.cil.oc.util.Reflection;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.util.text.ITextComponent;
import net.minecraftforge.common.util.Constants.NBT;
import net.minecraftforge.items.IItemHandler;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public final class AssemblerTemplates {
    public static final SlotInfo NoSlot = new SlotInfo(Slot.None(), Tier.None(), null, null);

    private static final List<Template> templates = new ArrayList<>();
    private static final List<ItemStackFilter> templateFilters = new ArrayList<>();

    public static void add(final NBTTagCompound template) {
        final ItemStackFilter selector = Reflection.getStaticMethod(template.getString("select"), ItemStackFilter.class);
        final TemplateValidator validator = Reflection.getStaticMethod(template.getString("validate"), TemplateValidator.class);
        final TemplateAssembler assembler = Reflection.getStaticMethod(template.getString("assemble"), TemplateAssembler.class);

        if (selector == null || validator == null || assembler == null) {
            return;
        }

        final Class<?> hostClass = Reflection.getClass(template.getString("hostClass"));
        final SlotInfo[] containerSlots = parseSlots(template.getTagList("containerSlots", NBT.TAG_COMPOUND), 3, Slot.Container(), hostClass);
        final SlotInfo[] upgradeSlots = parseSlots(template.getTagList("upgradeSlots", NBT.TAG_COMPOUND), 9, Slot.Upgrade(), hostClass);
        final SlotInfo[] componentSlots = parseSlots(template.getTagList("componentSlots", NBT.TAG_COMPOUND), 9, Slot.Any(), hostClass);

        templates.add(new Template(selector, validator, assembler, containerSlots, upgradeSlots, componentSlots));
    }

    public static void addFilter(final String method) {
        final ItemStackFilter filter = Reflection.getStaticMethod(method, ItemStackFilter.class);
        if (filter != null) {
            templateFilters.add(filter);
        }
    }

    @Nullable
    public static Template select(final ItemStack stack) {
        if (stack.isEmpty()) {
            return null;
        }

        for (final ItemStackFilter filter : templateFilters) {
            if (!filter.select(stack)) {
                return null;
            }
        }

        for (final Template template : templates) {
            if (template.select(stack)) {
                return template;
            }
        }

        return null;
    }

    // ----------------------------------------------------------------------- //

    public static final class Template {
        private final ItemStackFilter selector;
        private final TemplateValidator validator;
        private final TemplateAssembler assembler;
        public final SlotInfo[] containerSlots;
        public final SlotInfo[] upgradeSlots;
        public final SlotInfo[] componentSlots;

        Template(final ItemStackFilter selector, final TemplateValidator validator, final TemplateAssembler assembler, final SlotInfo[] containerSlots, final SlotInfo[] upgradeSlots, final SlotInfo[] componentSlots) {
            this.selector = selector;
            this.validator = validator;
            this.assembler = assembler;
            this.containerSlots = containerSlots;
            this.upgradeSlots = upgradeSlots;
            this.componentSlots = componentSlots;
        }

        public boolean select(final ItemStack stack) {
            return selector.select(stack);
        }

        public ValidationResult validate(final IItemHandler inventory) {
            final Object[] result = validator.validate(inventory);
            boolean isValid = false;
            ITextComponent progress = null;
            ITextComponent[] warnings = new ITextComponent[0];
            if (result.length > 0 && result[0] instanceof Boolean) {
                isValid = (boolean) result[0];
            }
            if (result.length > 1 && result[1] instanceof ITextComponent) {
                progress = (ITextComponent) result[1];
            }
            if (result.length > 2 && result[2] instanceof ITextComponent[]) {
                warnings = (ITextComponent[]) result[2];
            }
            return new ValidationResult(isValid, progress, warnings);
        }

        public AssemblyResult assemble(final IItemHandler inventory) {
            final Object[] result = assembler.assemble(inventory);
            ItemStack output = ItemStack.EMPTY;
            double requiredEnergy = 0;
            if (result.length > 0 && result[0] instanceof ItemStack) {
                output = (ItemStack) result[0];
            }
            if (result.length > 1 && result[1] instanceof Double) {
                requiredEnergy = (double) result[1];
            }
            return new AssemblyResult(output, requiredEnergy);
        }
    }

    public static final class SlotInfo {
        private final String type;
        private final int tier;
        private final SlotValidator validator;
        private final Class<?> hostClass;

        SlotInfo(final String type, final int tier, @Nullable final SlotValidator validator, @Nullable final Class<?> hostClass) {
            this.type = type;
            this.tier = tier;
            this.validator = validator;
            this.hostClass = hostClass;
        }

        public boolean validate(final IItemHandler inventory, final int slot, final ItemStack stack) {
            if (validator != null) {
                return validator.validate(inventory, slot, tier, stack);
            } else {
                final DriverItem driver;
                if (hostClass != null) {
                    driver = Driver.driverFor(stack, hostClass);
                } else {
                    driver = Driver.driverFor(stack);
                }

                return driver != null && Objects.equals(driver.slot(stack), type) && driver.tier(stack) <= tier;
            }
        }
    }

    public static final class ValidationResult {
        private final boolean isValid;
        private final ITextComponent progress;
        private final ITextComponent[] warnings;

        ValidationResult(final boolean isValid, @Nullable final ITextComponent progress, final ITextComponent[] warnings) {
            this.isValid = isValid;
            this.progress = progress;
            this.warnings = warnings;
        }

        public boolean isValid() {
            return isValid;
        }

        @Nullable
        public ITextComponent getProgress() {
            return progress;
        }

        public ITextComponent[] getWarnings() {
            return warnings;
        }
    }

    public static final class AssemblyResult {
        private final ItemStack output;
        private final double requiredEnergy;

        AssemblyResult(final ItemStack output, final double requiredEnergy) {
            this.output = output;
            this.requiredEnergy = requiredEnergy;
        }

        public ItemStack getOutput() {
            return output;
        }

        public double getRequiredEnergy() {
            return requiredEnergy;
        }
    }

    // ----------------------------------------------------------------------- //

    private static SlotInfo[] parseSlots(final NBTTagList slotsNbt, final int slotCount, @Nullable final String slotType, @Nullable final Class<?> hostClass) {
        final SlotInfo[] result = new SlotInfo[slotCount];
        Arrays.fill(result, NoSlot);
        for (int slot = 0; slot < Math.min(slotCount, slotsNbt.tagCount()); slot++) {
            final NBTTagCompound slotNbt = slotsNbt.getCompoundTagAt(slot);
            parseSlot(slotNbt, slotType, hostClass);
        }
        return result;
    }

    private static SlotInfo parseSlot(final NBTTagCompound slotNbt, @Nullable final String slotType, @Nullable final Class<?> hostClass) {
        final String type = slotType != null ? slotType : slotNbt.hasKey("type", NBT.TAG_STRING) ? slotNbt.getString("type") : Slot.None();
        final int tier = slotNbt.hasKey("tier", NBT.TAG_INT) ? slotNbt.getInteger("tier") : Tier.Any();
        final SlotValidator validator = slotNbt.hasKey("validate", NBT.TAG_STRING) ? Reflection.getStaticMethod(slotNbt.getString("validate"), SlotValidator.class) : null;
        return new SlotInfo(type, tier, validator, hostClass);
    }

    // ----------------------------------------------------------------------- //

    @FunctionalInterface
    private interface ItemStackFilter {
        boolean select(final ItemStack stack);
    }

    @FunctionalInterface
    private interface SlotValidator {
        boolean validate(final IItemHandler inventory, final int slot, final int tier, final ItemStack stack);
    }

    @FunctionalInterface
    private interface TemplateValidator {
        Object[] validate(final IItemHandler inventory);
    }

    @FunctionalInterface
    private interface TemplateAssembler {
        Object[] assemble(final IItemHandler inventory);
    }
}
